pytensor/tensor/rewriting/basic.py

```diff
@@ -135,6 +135,18 @@ def alloc_like(
     return rval
 
 
+def make_inplace(node, inplace_prop="inplace"):
+    op = getattr(node.op, "core_op", node.op)
```

Comment:
I missed this. Making the core_op inplace isn't enough because that falls out of scope of the pytensor safeguards. 

In these cases the Blockwise itself needs to be inplace as well. I forgot Cholesky was Blockwised by default now

Reply:
Need to just think a bit about the API for Blockwise inplace. I would restrict the rewrite to the core Op case now.

Reply:
When a Blockwise is not needed it will be removed so the inplace is still useful, just not for the batched case

Reply:
So I need to add a check to the rewrite that the Op is not a Blockwise?

Reply:
I'm a little confused, if you rewrite the test like this so `Blockwise` isn't compiled out:

```python
X = pt.tensor3("X")
L = pt.linalg.cholesky(X, overwrite_a=False, lower=True)
f = pytensor.function([pytensor.In(X, mutable=True)], L)
```

then the `overwrite_a` stays false so the rewrite wasn't applied, 

```python
pytensor.dprint(f)
# Blockwise{Cholesky{lower=True, overwrite_a=False, on_error='raise'}, (m,m)->(m,m)} [id A] 0
#└─ X [id B]
```

You want the rewrite to also happen in this case right, or no?

Reply:
Blockwise is only compiled out when not needed which is when there are no batch dims.

We want to support when it's needed, it's just that we need to think about that a bit better. For example we don't want the core op to inplace on an input that the Blockwise is broadcasting (doesn't apply to Cholesky which has a single input). This would corrupt the input for subsequent iterations of the core op call.

Also there's two levels of inplace. The one the core op can do and the one the Blockwise can do regardless of how the core op is implemented. 

The Blockwise sends a subset of the inputs to the core op in a loop, and then it can decide to overwrite the input with the intermediate results even if the core op has no inplacing of its own.

The Blockwise level inplace can be implemented automatically from inspecting the signature and seeing if any input has the same core shape as the output(s) (like  "`(m,n),...->(m,n)`") and is not being broadcasted. 

Even better would be to try and align this with the lower level inplacing of the core op. This logic should happen in the rewrite but it's a tad bit more complex than what is written now.

I was suggesting we handle it in a separate PR but we can do it here. Intermediately, we can also restrict inplace to single input output Blockwise for now which is a tad simpler.

Reply:
> For example we don't want the core op to inplace on an input that the Blockwise is broadcasting (doesn't apply to Cholesky which has a single input)

Gotcha, that makes sense.  Though, I'm assuming this PR is focused on the stuff in slinalg, where I'm having a hard time thinking of a case when there'd be some broadcasting happening.  Do you know of an example?  Most of these seem in the same vein as Cholesky, where Blockwise will repeat the decompositon or solve on each matrix in the stacked input, and can overwrite without causing problems.  

I'll poke around more in the codebase and see where Blockwise is used.  I think I follow the rest of what you're saying but I need to poke at some concrete examples to grok it better.

Reply:
Solve could broadcast the A matrix with a batch of b vectors. Anything with more than 1 input can be broadcasting in Blockwise (whether that ever happens in practice for specific Ops is a different question)

Reply:
Why wouldn't a blockwise solve(A, b, overwrite_a=True) be safe?

Reply:
Yeah what matters is not the batching but the broadcasting (which is caused by the batching of other inputs)

Since Blockwise is more general than linalg Ops it may be worth to think about a couple more Ops to see if there's a strategy that works reasonably well for most cases.

Also may be worth thinking a bit about the numba backend.

Reply:
>  think it boils down to "overwrite things that are already batched, refuse to overwrite if it's being broadcast"

Yeah and for that we need a minimal API so Blockwise can talk with the core Op in a sense.

I think it's enough for Blockwise to say what inputs are destructible and the Core Op can decide (but is not required) to destroy those, but none else.

I am hopeful that may not require too much code.

Reply:
So basically what needs to happen is for blockwise to get some kind of "safe_to_destroy_map" property that can be referenced by rewrites?

Reply:
Do these issues only come up when `Blockwise` isn't compiled out?  Would it be possible to punt on it till later, or make it a separate issue, and go with Jesses current approach of applying the rewrite on the `core_op`?  If inplacing only worked on non-batched matrices I think that would satisfy the pareto principle pretty well.